apps/src/JSInterpreter.js

```diff
@@ -78,25 +86,69 @@ JSInterpreter.prototype.patchInterpreterMethods_ = function () {
     this.throwException('Unknown identifier: ' + nameStr);
     return this.UNDEFINED;
   };
+
   /**
   * Sets a value to the current scope.
   * @param {!Object} name Name of variable.
   * @param {!Object} value Value.
+  * @param {boolean} declarator true if called from variable declarator.
   */
-  window.Interpreter.prototype.setValueToScope = function (name, value) {
+  window.Interpreter.prototype.setValueToScope = function (name, value, declarator) {
     var scope = this.getScope();
     var strict = scope.strict;
     var nameStr = name.toString();
     while (scope) {
       if (this.hasProperty(scope, nameStr) || (!strict && !scope.parentScope)) {
-        this.setProperty(scope, nameStr, value);
+        if (declarator) {
+          // from a declarator, always call baseSetProperty
+          JSInterpreter.baseSetProperty.call(this, scope, nameStr, value);
+        } else {
+          this.setProperty(scope, nameStr, value);
+        }
         return;
       }
       scope = scope.parentScope;
     }
     this.throwException('Unknown identifier: ' + nameStr);
   };
 
+  /**
+   * Sets a value to the scope chain or to an object property.
+   * @param {!Object|!Array} left Name of variable or object/propname tuple.
+   * @param {!Object} value Value.
+   * @param {boolean} declarator true if called from variable declarator.
+   */
+  window.Interpreter.prototype.setValue = function (left, value, declarator) {
+    if (left.length) {
+      var obj = left[0];
+      var prop = left[1];
+      this.setProperty(obj, prop, value);
+    } else {
+      this.setValueToScope(left, value, declarator);
+    }
+  };
+
+  // Patched to add the 3rd "declarator" parameter on the setValue() call(s).
+  // Also removed erroneous? call to hasProperty when there is node.init
+  // Changed to call setValue with this.UNDEFINED when there is no node.init
+  // and JSInterpreter.baseHasProperty returns false for current scope.
+  window.Interpreter.prototype['stepVariableDeclarator'] = function () {
```

Comment:
This is the way it is written in the interpreter.js version - I'm just copying to avoid introducing extra diffs for managing future merges into this copied code.